joy_teleop/test/test_action_fibonacci.py

```diff
@@ -0,0 +1,136 @@
+    parameters['fibonacci.action_name'] = '/fibonacci'
+    parameters['fibonacci.buttons'] = [2]
+    parameters['fibonacci.action_goal'] = {'order': 5}
+    joy_teleop_node = launch_ros.actions.Node(
+        package='joy_teleop',
+        node_executable='joy_teleop',
+        output='both',
+        parameters=[parameters])
+
+    return (
+        launch.LaunchDescription([
+            joy_teleop_node,
+            launch_testing.actions.ReadyToTest(),
+        ]),
+        {
+            'joy_teleop_node': joy_teleop_node,
+        }
+    )
+
+
+class TestJoyTeleopActionFibonacci(unittest.TestCase):
+
+    def setUp(self):
+        self.context = rclpy.context.Context()
+        rclpy.init(context=self.context)
+        self.node = rclpy.create_node('test_action_fibonacci', context=self.context)
+        self.executor = rclpy.executors.SingleThreadedExecutor(context=self.context)
+        self.executor.add_node(self.node)
+
+    def tearDown(self):
+        self.node.destroy_node()
+        rclpy.shutdown(context=self.context)
+
+    def test_simple_message(self, launch_service, joy_teleop_node, proc_output):
+        joy_publisher = self.node.create_publisher(
+            msg_type=sensor_msgs.msg.Joy,
+            topic='joy',
+            qos_profile=1,
+        )
+
+        sequence = []
+        future = rclpy.task.Future()
+
+        def fibonacci_callback(goal_handle):
+            nonlocal future
+
+            sequence.append(0)
+            sequence.append(1)
+            for i in range(1, goal_handle.request.order):
+                sequence.append(sequence[i] + sequence[i-1])
+
+            goal_handle.succeed()
+            result = action_tutorials_interfaces.action.Fibonacci.Result()
+            result.sequence = sequence
+            future.set_result(True)
+            return result
+
+        action_server = rclpy.action.ActionServer(
+            self.node,
+            action_tutorials_interfaces.action.Fibonacci,
+            'fibonacci',
+            fibonacci_callback)
+
+        joy_msg = sensor_msgs.msg.Joy()
+        joy_msg.header.stamp.sec = 0
+        joy_msg.header.stamp.nanosec = 0
+        joy_msg.header.frame_id = ''
+        joy_msg.axes = []
+        # Above we set the button to be used as '2', so here we set the '2' button active.
+        joy_msg.buttons = [0, 0, 1]
+
+        def publish_message():
+            joy_publisher.publish(joy_msg)
+            joy_msg.buttons[2] = int(not joy_msg.buttons[2])
+        publish_timer = self.node.create_timer(0.1, publish_message)
+
+        self.executor.spin_until_future_complete(future, timeout_sec=10)
+
+        # Check
+        assert future.done() and future.result(), 'Timed out waiting for action to complete'
+        self.assertTrue(sequence == [0, 1, 1, 2, 3, 5])
+
+        # Cleanup
+        action_server.destroy()
+        self.node.destroy_timer(publish_timer)
+        joy_publisher.destroy()
```

Comment:
@clalancette nit: here and elsewhere, this may never happen if an previous assertion fails.

Reply:
Hm, true.

The truly correct way to do this would be something like:

```
pub = create_publisher()
try:
    timer = create_timer()
    try:
        action_server = ActionServer()
        try:
          # Do test stuff in here
        finally:
          action_server.destroy()
    finally:
        destroy_timer(timer)
finally:
    pub.destroy()
```

But it is pretty wordy.  Other thoughts on how to do this nicely?

Reply:
Never mind this one, actually.  After the other revamps you suggested, this became a lot easier.